### PR TITLE
Fix startup focus state potentially being incorrect on Windows

### PR DESCRIPTION
**Pull request by peppy**
## Description

This is a well documented flaw at windows' end, where a window will not receive an initial WM_ACTIVATE event if the window is created while another window has focus. Nor will it receive the (first) subsequent WM_ACTIVATE *after* that, making initial focus detection on Windows quite a pain to handle.

Two changes are required here to make this work:

- Double check whether the SDL window has focus rather than blindly accepting that WM_ACTIVATE guaranteed focus.
- Also check focus state on WM_PAINT, to handle the missing subsequent WM_ACTIVATE event.

Tested only on Windows 10, but this should not cause any backwards compatibility issues.

## Testing

A bit of a manual one, by dragging another window while starting an SDL application.

I've been using the following code, with a thread sleep to give me time to drag another window, then monitoring SDL event flow for `SDL_WINDOWEVENT_FOCUS_GAINED` and `SDL_WINDOWEVENT_FOCUS_LOST` to arrive correctly.

Before merging, I'd appreciate if someone with more knowledge of SDL can confirm whether leaving the `keyboard->focus` flag in its prior state is what we want. With my changes, I'm fixing the flow of events, but there's still a potential that `SDL_Keyboard` is seeing an incorrect focus value.

It shouldn't be a huge stretch to refactor to also update the keyboard's focus, if that's what is deemed more correct.

```csharp

using System;
using System.Threading;
using SDL2;

namespace SDL2IsolatedRelativeModeTest
{
    class Program
    {
        static void Main(string[] args)
        {
            SDL.SDL_Init(SDL.SDL_INIT_VIDEO);
            SDL.SDL_SetHint(SDL.SDL_HINT_EVENT_LOGGING, "2");

            Thread.Sleep(5000);

            var win1 = SDL.SDL_CreateWindow("test window", 100, 100,500, 500, SDL.SDL_WindowFlags.SDL_WINDOW_RESIZABLE);

            while (true)
            {
                while (SDL.SDL_PollEvent(out var e) > 0)
                {
                    switch (e.type)
                    {
                        case SDL.SDL_EventType.SDL_WINDOWEVENT:
                            switch (e.window.windowEvent)
                            {
                                case SDL.SDL_WindowEventID.SDL_WINDOWEVENT_RESTORED:
                                case SDL.SDL_WindowEventID.SDL_WINDOWEVENT_FOCUS_GAINED:
                                    Console.WriteLine($"{e.window.windowID} Got focus gained");
                                    break;

                                case SDL.SDL_WindowEventID.SDL_WINDOWEVENT_MINIMIZED:
                                case SDL.SDL_WindowEventID.SDL_WINDOWEVENT_FOCUS_LOST:
                                    Console.WriteLine($"{e.window.windowID} Got focus lost");
                                    break;
                            }

                            break;
                    }

                    break;
                }
            }
        }
    }
}


```

## Existing Issue(s)

Closes https://github.com/libsdl-org/SDL/issues/3076.

### CODE REVIEW

File: src/video/windows/SDL_windowsevents.c [hunk 1]
```diff
@@ -490,6 +490,22 @@ WIN_KeyboardHookProc(int nCode, WPARAM wParam, LPARAM lParam)
     return 1;
 }
 
+static void UpdateFocus(HWND hwnd, SDL_Window * window)
+{
+    if (SDL_GetKeyboardFocus() != window) {
```
Comment (peppy) [inline src/video/windows/SDL_windowsevents.c:495]: In line with my query in the PR description, I'd like confirmation whether this check should be here. The alternative is reversing the order of checks and updating `SetKeyboardFocus` with the corrected values.

**Comment by DomGries** [comment]
I can confirm that this fixes the long standing bug of incorrect input focus when the SDL app is started in the background.

This can be reproduced easily for any SDL app by clicking on the exe in explorer and then quickly clicking on another file in explorer before the app managed to start.

In case of using relative mouse movement at the start this then also locks the mouse to the center of the screen even while the SDL app is unfocused which is problematic.

As such it would be great if this can be looked at and merged in time for the next release.

@peppy Do I assume correctly that WM_PAINT is only called once and not for each frame?

**Comment by peppy** [comment]
@DomGries I did notice it firing more than once (I think on focus switch or restore). Don't recall the specifics but it's safe to say it doesn't cause any kind of performance issue. Definitely not on every update.

**Comment by peppy** [comment]
Heads up that since deploying this we have had a [couple of reports](https://github.com/ppy/osu/issues/12644) of the initial `focused=true` state never arriving, even though the user does have the window focused. Not sure on a solution to this yet. May have to poll for state or something equally ugly.

**Comment by slouken** [comment]
Can you check to see if https://github.com/libsdl-org/SDL/commit/632aca29453cf09a44bd45497363607a106634f8 fixes this issue?

**Comment by peppy** [comment]
> Can you check to see if [632aca2](https://github.com/libsdl-org/SDL/commit/632aca29453cf09a44bd45497363607a106634f8) fixes this issue?

Unfortunately it does not fix the issue. The `SDL_WINDOWEVENT_FOCUS_GAINED` event is still unexpectedly arriving on startup. The test code I provided still works well to 100% repro this.

**Comment by slouken** [comment]
> > Can you check to see if [632aca2](https://github.com/libsdl-org/SDL/commit/632aca29453cf09a44bd45497363607a106634f8) fixes this issue?
> 
> Unfortunately it does not fix the issue. The `SDL_WINDOWEVENT_FOCUS_GAINED` event is still unexpectedly arriving on startup. The test code I provided still works well to 100% repro this.

With the latest code, I haven't found a case where it's out of sync. What are your repro steps?

Here are my steps:
1. Run the test program
2. Drag another window by the title bar during the SDL_Delay() call
3. See the SDL window pop up and get input focus, note that the window drag stopped and key events go to the SDL window.

Are you doing something different?

Here's the sample code converted to C, for reference:
```
#include "SDL.h"

int main( int argc, char *argv[] )
{
	SDL_Init( SDL_INIT_VIDEO );
	SDL_SetHint( SDL_HINT_EVENT_LOGGING, "2" );

	SDL_Delay( 5000 );

	SDL_CreateWindow( "test window", 100, 100, 500, 500, SDL_WINDOW_RESIZABLE );

	while ( 1 )
	{
		SDL_Event e;
		while ( SDL_PollEvent( &e ) > 0 )
		{
			switch ( e.type )
			{
			case SDL_WINDOWEVENT:
				switch ( e.window.event )
				{
				case SDL_WINDOWEVENT_RESTORED:
				case SDL_WINDOWEVENT_FOCUS_GAINED:
					SDL_Log( "Got focus gained\n" );
					break;

				case SDL_WINDOWEVENT_MINIMIZED:
				case SDL_WINDOWEVENT_FOCUS_LOST:
					SDL_Log( "Got focus lost\n" );
					break;
				}

				break;
			}

			break;
		}
	}
}
```

**Comment by DomGries** [comment]
For me 632aca2 also does not fix the issue.

> 3. See the SDL window pop up and get input focus [...]

When reproducing correctly Windows will block the window from getting to the foreground which causes the issue.

Please make sure to not start the test program from Visual Studio but rather from explorer. Also please finish the drag of another window within the delay timespan or simply just repeatedly click on another window (e.g. the explorer window where you started the program from).

**Comment by slouken** [comment]
You're right, in this case the test app doesn't get activated and get keyboard focus even though all the Windows messages indicate that it does:
```
[15672] Received windows message: 80e62 WM_WINDOWPOSCHANGING -- 0x0, 0x58F82C
[15672] Received windows message: 80e62 WM_NCCALCSIZE -- 0x1, 0x58F804
[15672] Received windows message: 80e62 WM_WINDOWPOSCHANGED -- 0x0, 0x58F82C
[15672] Received windows message: 80e62 WM_GETICON -- 0x2, 0x0
[15672] Received windows message: 80e62 WM_GETICON -- 0x0, 0x0
[15672] Received windows message: 80e62 WM_GETICON -- 0x1, 0x0
[15672] Received windows message: 80e62 WM_SETTEXT -- 0x0, 0xBD3338
[15672] Received windows message: 80e62 WM_STYLECHANGING -- 0xFFFFFFEC, 0x58F804
[15672] Received windows message: 80e62 WM_STYLECHANGED -- 0xFFFFFFEC, 0x58F804
[15672] Received windows message: 80e62 WM_SHOWWINDOW -- 0x1, 0x0
[15672] INFO: SDL EVENT: SDL_WINDOWEVENT (timestamp=5012 windowid=1 event=SDL_WINDOWEVENT_SHOWN data1=0 data2=0) 
[15672] Received windows message: 80e62 WM_WINDOWPOSCHANGING -- 0x0, 0x58F82C
[15672] Received windows message: 80e62 WM_WINDOWPOSCHANGING -- 0x0, 0x58F82C
[15672] Received windows message: 80e62 WM_ACTIVATEAPP -- 0x1, 0x0
[15672] Received windows message: 80e62 WM_NCACTIVATE -- 0x0, 0x0
[15672] Received windows message: 80e62 WM_ACTIVATE -- 0x1, 0x0
[15672] Received windows message: 80e62 WM_IME_SETCONTEXT -- 0x1, 0xC000000F
[15672] Received windows message: 80e62 WM_IME_NOTIFY -- 0x2, 0x0
[15672] Received windows message: 80e62 WM_GETOBJECT -- 0xFFFFFFFF, 0xFFFFFFFC
[15672] Received windows message: 80e62 WM_GETOBJECT -- 0x0, 0xFFFFFFF4
[15672] Received windows message: 80e62 WM_SETFOCUS -- 0x0, 0x0
[15672] INFO: SDL EVENT: SDL_WINDOWEVENT (timestamp=5023 windowid=1 event=SDL_WINDOWEVENT_FOCUS_GAINED data1=0 data2=0) 
[15672] INFO: SDL EVENT: SDL_TEXTEDITING (timestamp=5024 windowid=1 text='' start=0 length=0) 
[15672] Received windows message: 80e62 WM_IME_SETCONTEXT -- 0x0, 0xC000000F
[15672] Received windows message: 80e62 WM_IME_SETCONTEXT -- 0x1, 0xC000000F
[15672] Received windows message: 80e62 WM_IME_SETCONTEXT -- 0x0, 0xC000000F
[15672] Received windows message: 80e62 WM_IME_SETCONTEXT -- 0x1, 0xC000000F
[15672] Received windows message: 80e62 WM_NCPAINT -- 0x1, 0x0
[15672] Received windows message: 80e62 WM_ERASEBKGND -- 0x7C010617, 0x0
[15672] Received windows message: 80e62 WM_WINDOWPOSCHANGED -- 0x0, 0x58F82C
[15672] Received windows message: 80e62 WM_SIZE -- 0x0, 0x1F401F4
[15672] Received windows message: 80e62 WM_MOVE -- 0x0, 0x640064
[15672] Received windows message: 80e62 WM_GETICON -- 0x1, 0x60
[15672] Received windows message: 80e62 UNKNOWN (49475) -- 0x0, 0x0
[15672] Received windows message: 80e62 WM_GETICON -- 0x2, 0x60
[15672] Received windows message: 80e62 WM_GETMINMAXINFO -- 0x0, 0x58F7FC
[15672] Received windows message: 80e62 WM_GETICON -- 0x0, 0x60
[15672] Received windows message: 80e62 WM_PAINT -- 0x0, 0x0
[15672] INFO: SDL EVENT: SDL_WINDOWEVENT (timestamp=5026 windowid=1 event=SDL_WINDOWEVENT_EXPOSED data1=0 data2=0) 
[15672] INFO: Got focus gained 
```

**Comment by slouken** [comment]
Does https://github.com/libsdl-org/SDL/commit/c542de92135a79f25648e75b1c2ab0fd581e29c0 help?

**Comment by peppy** [comment]
> Does [c542de9](https://github.com/libsdl-org/SDL/commit/c542de92135a79f25648e75b1c2ab0fd581e29c0) help?

I haven't checked on the specifics of the code change, but this does look to fix the issue!

**Comment by DomGries** [comment]
Yes it also works for me now so #3076, #4293 and #4450 could probably be closed now. Thanks!

**Pull request closed without merging**